src/ctrace/src/ctrace.erl

```diff
@@ -0,0 +1,425 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(ctrace).
+
+-vsn(1).
+
+-behaviour(config_listener).
+
+-export([
+    with_span/2,
+    with_span/3,
+    start_span/1,
+    start_span/2,
+    finish_span/0,
+    finish_span/1,
+    update_tracers/0,
+    has_span/0,
+    external_span/3,
+
+    tag/1,
+    log/1,
+
+    tags/0,
+    refs/0,
+    operation_name/0,
+    trace_id/0,
+    span_id/0,
+    tracer/0,
+    context/0,
+
+    match/2,
+    filter_module_name/1
+]).
+
+-export([
+    handle_config_change/5,
+    handle_config_terminate/3
+]).
+
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("passage/include/opentracing.hrl").
+
+-define(MAIN_TRACER, jaeger_passage_reporter).
+-define(IS_ENABLED_KEY, ctrace_is_enabled).
+-define(ORIGIN_KEY, ctrace_origin_key).
+-define(ACTIONS_KEY, ctrace_actions_key).
+-define(ANCESTORS_KEY, passage_span_ancestors).
+
+-type tags()
+    :: #{atom() => term()}.
+
+-type log_fields()
+    :: #{atom() => term()}.
+
+-type start_span_options()
+    :: [start_span_option()].
+
+-type start_span_option()
+    :: {time, erlang:timespan()}
+    | {tags, tags()}.
+
+-type finish_span_options()
+    :: [finish_span_option()].
+
+-type finish_span_option()
+    :: {time, erlang:timespan()}.
+
+%% @equiv with_span(OperationName, [], Fun)
+-spec with_span(
+       OperationId :: atom(),
+       Fun
+   ) -> Result when
+       Fun :: fun (() -> Result),
+       Result :: term().
+
+with_span(OperationId, Fun) ->
+    with_span(OperationId, #{}, Fun).
+
+-spec with_span(
+       OperationId :: atom(),
+       TagsOrOptions :: tags() | start_span_options(),
+       Fun
+   ) -> Result when
+       Fun :: fun (() -> Result),
+       Result :: term().
+
+with_span(OperationId, ExtraTags, Fun) when is_map(ExtraTags) ->
+    with_span(OperationId, [{tags, ExtraTags}], Fun);
+
+with_span(OperationId, Options, Fun)  ->
+    try
+        start_span(OperationId, Options),
+        Fun()
+    after
```

Comment:
Ops, just noticed that we should log an error on exception like `passage_pd:with_span` does by default:

https://github.com/apache/couchdb/compare/prototype/fdb-layer...opentracing-davisp#diff-89ac3e12a9462a9e6c44129ae34d3a0eR122-R128